### PR TITLE
[3.11] Pod Retention for Jenkins Pipeline Builds

### PR DESCRIPTION
**Pull request by adambkaplan**
Doc for the Pod Retention setting added to the Jenkins
Kubernetes plugin.

Bug [151940](https://bugzilla.redhat.com/show_bug.cgi?id=1515940)

See related PR jenkinsci/kubernetes-plugin#354

### CODE REVIEW

**Comment by adambkaplan** [comment]
/hold
Wait until approval upstream

File: using_images/other_images/jenkins.adoc [hunk 1]
```diff
@@ -766,6 +766,47 @@ You can increase the amount of memory available to Jenkins by overriding
 the *MEMORY_LIMIT* paramenter when instantiating the Jenkins Ephemeral or
 Jenkins Persistent template.
 
+[[build-pod-retention]]
+=== Build Pod Retention
+
+Pipeline build pods are deleted by default after the build completes or is aborted.
```
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: I would clarify that these build pods are Jenkins agent/slave pods created by kubernetes plugin.

**Review by gabemontero — CHANGES_REQUESTED** [review]
I would move this to using_images/other_images/jenkins_slave.adoc

**Comment by adambkaplan** [comment]
@gabemontero I'm a bit less inclined to move this section to the `jenkins_slave` page, since the latter goes into the details of our Maven and node.js slave images. Furthermore the behavior is governed by the kube plugin, which "lives" on the master.

**Comment by gabemontero** [comment]
@adambkaplan ... still mulling your rationale on the page location, but in the interim,  if we don't move the content, I think a link to your new section here on that page (perhaps in the beginning of the usage section) is helpful.  Your thoughts there?

**Comment by adambkaplan** [comment]
@gabemontero moved it over with an "intro note" on the jenkins page.

**Review by gabemontero — APPROVED** [review]
cool ... thanks @adambkaplan

**Comment by vikram-redhat** [comment]
@adambkaplan - is this ok to move forward with?

**Comment by adambkaplan** [comment]
@vikram-redhat this was merged upstream yesterday. Currently waiting on a new version of the plugin to be released, then we can add it to our Jenkins image.

**Comment by vikram-redhat** [comment]
Thanks @adambkaplan - ping the openshift docs team when ready.

File: using_images/other_images/jenkins.adoc [hunk 2]
```diff
@@ -727,6 +727,11 @@ specified.
 <9> The node stanza references the name of the Pod template newly defined above.
 ====
 
+By default the pod is deleted when the build completes.
+This behavior can be modified via the plug-in or within a pipeline Jenkinsfile - see 
+xref:../../using_images/other_images/jenkins_slaves.adoc#slave-pod-retention[Slave Pod Retention]
```
Comment (bmcelvee) [inline using_images/other_images/jenkins.adoc]: Are we still referring to these as Jenkins "Slaves" or can it be changed to "Agents"?

File: using_images/other_images/jenkins_slaves.adoc [hunk 3]
```diff
@@ -236,3 +237,44 @@ variables.
 * Set the maximum heap size and JVM arguments for any Gradle test JVM by
 the maxHeapSize and jvmArgs settings in build.gradle, or though the
 `-Dorg.gradle.jvmargs` command line argument.
+
+[[slave-pod-retention]]
+== Slave Pod Retention
+
+Jenkins slave pods are deleted by default after the build completes or is aborted. 
+This behavior can be changed via the Kubernetes plug-in "Pod Retention" setting.
```
Comment (bmcelvee) [inline using_images/other_images/jenkins_slaves.adoc]: s/via/by
s/"Pod Retention"/*Pod Retention*

File: using_images/other_images/jenkins_slaves.adoc [hunk 4]
```diff
@@ -236,3 +237,44 @@ variables.
...
+
+[[slave-pod-retention]]
+== Slave Pod Retention
+
+Jenkins slave pods are deleted by default after the build completes or is aborted. 
+This behavior can be changed via the Kubernetes plug-in "Pod Retention" setting. 
+Pod retention can be set for all Jenkins builds, with overrides for each 
+pod template. The following behaviors are supported:
+
+- "Always" will keep the build pod regardless of build result.
```
Comment (bmcelvee) [inline using_images/other_images/jenkins_slaves.adoc]: s/"..."/`...` (for all of the list items)
s/will keep/keeps
s/will use/uses

File: using_images/other_images/jenkins_slaves.adoc [hunk 5]
```diff
@@ -236,3 +237,44 @@ variables.
...
+- "Default" will use the plug-in value (pod template only).
+- "Never" always deletes the pod.
+- "On Failure" keeps the pod if it fails during the build.
+
+Pod retention can also be overrode in the pipeline Jenkinsfile as follows:
```
Comment (bmcelvee) [inline using_images/other_images/jenkins_slaves.adoc]: s/Pod retention can also be overrode in the pipeline Jenkinsfile as follow:/You can override pod retention in the pipeline Jenkinsfile:

File: using_images/other_images/jenkins_slaves.adoc [hunk 6]
```diff
@@ -236,3 +237,44 @@ variables.
...
+
+====
```
Comment (bmcelvee) [inline using_images/other_images/jenkins_slaves.adoc]: We can remove the ==== here (puts a box around the code block)

File: using_images/other_images/jenkins_slaves.adoc [hunk 7]
```diff
@@ -236,3 +237,44 @@ variables.
...
+  containers: [
+    ...
+  ]) {
+  node("mypod") {
+    ...
+  }
+}
+----
+<1> Allowed values for `podRetention` are `never()`, `onFailure()`, `always()`, and `default()`.
+====
```
Comment (bmcelvee) [inline using_images/other_images/jenkins_slaves.adoc]: We can remove the ==== here (puts a box around the code block)

File: using_images/other_images/jenkins_slaves.adoc [hunk 8]
```diff
@@ -236,3 +237,44 @@ variables.
...
+
+[WARNING]
+====
+Pods that are kept may continue to run and count against resource quotas.
+====
+
+See the link:https://github.com/jenkinsci/kubernetes-plugin[Kubernetes plug-in
```
Comment (bmcelvee) [inline using_images/other_images/jenkins_slaves.adoc]: ifdef::openshift-origin[]
For more information, see the link:https://github.com/jenkinsci/kubernetes-plugin[Kubernetes plug-in documentation].
endif::openshift-origin[]

**Comment by bmcelvee** [comment]
Just a few nits. Otherwise LGTM :stars: 

When it's ready we'll do QE review too.

File: using_images/other_images/jenkins.adoc (hunk 2 above)
Comment (adambkaplan) [reply using_images/other_images/jenkins.adoc]: Looks like "Agents" is clearer - I can change it

**Comment by adambkaplan** [comment]
/hold cancel

per [openshift/jenkins#672](https://github.com/openshift/jenkins/pull/672#issuecomment-409673498) this is released.

**Comment by bmcelvee** [comment]
@xltian PTAL

**Comment by xltian** [comment]
@wzheng1 as the QE owner for this feature will review.

**Comment by bmcelvee** [comment]
Thanks, @xltian and @wzheng1!

**Comment by adambkaplan** [comment]
@wzheng1 should have updated the Trello card. This was completed Tuesday and should be available in the latest openshift Jenkins image.

**Comment by wzheng1** [comment]
Feature is still under testing and with bug blocked. But it look good to me just for this doc.

**Comment by bmcelvee** [comment]
Thanks, @bparees!

**Pull request merged**